README.rst
LWEncryptionMachine
===================

.. image:: https://img.shields.io/badge/license-MIT-yellow.svg
    :target: LICENSE.txt
.. image:: https://img.shields.io/github/v/release/leeway64/LWEncryptionMachine
   :alt: GitHub release (latest by date)

LWEncryptionMachine is a simple encryption machine that can encrypt a user-entered cryptographic key
and message using a Caesar cipher.


Installation
--------------

.. code-block::

    git clone https://github.com/leeway64/LWEncryptionMachine.git
    cd LWEncryptionMachine
    make build-all


An introduction to ciphers
---------------------------

This project uses a Caesar cipher; more information on Caesar ciphers can be found
`here <https://en.wikipedia.org/wiki/Caesar_cipher>`_.

In general, ciphers work by transforming plaintext into ciphertext through a cipher.

.. image:: doc/cipher-encryption-process.png

The previous diagram is an `activity diagram <https://plantuml.com/activity-diagram-beta>`_ drawn
with `PlantUML <https://plantuml.com/>`_. To render the diagram, run the following commands:

.. code-block::

    python3 -m venv .venv
    source .venv/bin/activate
    pip install -r plantuml_requirements.txt
    python -m plantuml doc/cipher-encryption-process.puml


Examples
----------

.. code-block::

        $ bin/LWEncryptionMachine

        Incorrect number of arguments. Enter the alphabet and the number of letters to be shifted by.
        For example: LWEncryptionMachine abcdefghijklmnopqrstuvwxyz0123456789 3


        $ bin/LWEncryptionMachine abcdefghijklmnopqrstuvwxyz0123456789 3

        Welcome to LWEncryptionMachine
        The program lets you encrypt a message
        with a key for your recipient to decrypt!

        Alphabet entered: abcdefghijklmnopqrstuvwxyz0123456789
        Shift value entered: 3

        Encrypted messages use a shared keyword to decrypt.
          Enter key: Justice
        "Justice" has been encrypted to: cxvwlfh

        How many words is your message? 2
          Next word: free
        "free" has been encrypted to: iuhh
          Next word: tibet
        "tibet" has been encrypted to: wlehw

        Message fully encrypted. Happy secret messaging!


        $ bin/LWEncryptionMachine ABCDEFGHIJKLMNOPQRSTUVWXYZ 7

        Welcome to LWEncryptionMachine
        The program lets you encrypt a message
        with a key for your recipient to decrypt!

        Alphabet entered: ABCDEFGHIJKLMNOPQRSTUVWXYZ
        Shift value entered: 7

        Encrypted messages use a shared keyword to decrypt.
          Enter key: FREEDOM
            "FREEDOM" has been encrypted to: MYLLKVT

        How many words is your message? 4
          Next word: EAST
            "EAST" has been encrypted to: LHZA
          Next word: TURKISTAN
            "TURKISTAN" has been encrypted to: ABYRPZAHU
          Next word: HONG
            "HONG" has been encrypted to: OVUN
          Next word: KONG
            "KONG" has been encrypted to: RVUN

        Message fully encrypted. Happy secret messaging!


Running unit tests
-------------------

.. code-block::

    make run-tests

Some unit tests were influenced by certain examples given in the files section of
`Assessment 3 <https://courses.cs.washington.edu/courses/cse142/21su/assessments/a3/>`_
for `UW Seattle CSE 142 (summer 2021) <https://courses.cs.washington.edu/courses/cse142/21su/>`_.


Acknowledgements
------------------

This project is based on `Assessment 3 <https://courses.cs.washington.edu/courses/cse142/21su/assessments/a3/>`_
for `UW Seattle CSE 142 (summer 2021) <https://courses.cs.washington.edu/courses/cse142/21su/>`_.
The assignment was originally meant to be completed in Java, but I completed it in C.


Third-party software
----------------------

- `CMake <https://cmake.org/>`_ (BSD 3-clause License): C/C++ build system.
- `CuTest <http://cutest.sourceforge.net/>`_ (zlib License): C unit testing framework.
- `Python-PlantUML <https://github.com/dougn/python-plantuml/>`_ (BSD-2-Clause license): Python package for generating PlantUML diagrams.


CHANGELOG.md
# Changelog

All changes between versions for this project will be documented in this file.

This changelog format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
this project follows [semantic versioning](https://semver.org/).


## Unreleased
- None


## [1.0.1] - 2023-6-10
### Changed
- Updated README
 

## [1.0.0] - 2023-5-13
- Initial release


test/test_suite.c
#include <stdio.h>
#include "CuTest.h"

#include "../src/LWEncryptionMachine.h"


void testIndexOf(CuTest* tc) {
    CuAssertIntEquals(tc, indexOf('a', "abcdefghijklmnopqrstuvwxyz"), 0);
    CuAssertIntEquals(tc, indexOf('b', "abcdefghijklmnopqrstuvwxyz"), 1);
    CuAssertIntEquals(tc, indexOf('c', "abcdefghijklmnopqrstuvwxyz"), 2);
    CuAssertIntEquals(tc, indexOf('s', "abcdefghijklmnopqrstuvwxyz"), 18);
    CuAssertIntEquals(tc, indexOf('x', "abcdefghijklmnopqrstuvwxyz"), 23);
    CuAssertIntEquals(tc, indexOf('y', "abcdefghijklmnopqrstuvwxyz"), 24);
    CuAssertIntEquals(tc, indexOf('z', "abcdefghijklmnopqrstuvwxyz"), 25);
    
    CuAssertIntEquals(tc, indexOf('0', "abcdefghijklmnopqrstuvwxyz0123456789"), 26);
    CuAssertIntEquals(tc, indexOf('2', "abcdefghijklmnopqrstuvwxyz0123456789"), 28);
    CuAssertIntEquals(tc, indexOf('4', "abcdefghijklmnopqrstuvwxyz0123456789"), 30);
    CuAssertIntEquals(tc, indexOf('8', "abcdefghijklmnopqrstuvwxyz0123456789"), 34);
                    
}

void testEncryptLetter(CuTest* tc) {
    CuAssertIntEquals(tc, encryptLetter('a', "abcdefghijklmnopqrstuvwxyz", 0), 'a');
    CuAssertIntEquals(tc, encryptLetter('d', "abcdefghijklmnopqrstuvwxyz", 1), 'e');
    CuAssertIntEquals(tc, encryptLetter('e', "abcdefghijklmnopqrstuvwxyz", 2), 'g');
    CuAssertIntEquals(tc, encryptLetter('f', "abcdefghijklmnopqrstuvwxyz", 3), 'i');
    
    CuAssertIntEquals(tc, encryptLetter('n', "abcdefghijklmnopqrstuvwxyz", 4), 'r');
    CuAssertIntEquals(tc, encryptLetter('y', "abcdefghijklmnopqrstuvwxyz", 5), 'd');
    CuAssertIntEquals(tc, encryptLetter('z', "abcdefghijklmnopqrstuvwxyz", 6), 'f');

    CuAssertIntEquals(tc, encryptLetter('0', "0123456789", 3), '3');
    CuAssertIntEquals(tc, encryptLetter('8', "0123456789", 4), '2');

}

void testEncryptWord(CuTest* tc) {
    char str1[] = "play";
    encryptWord(str1, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str1, "sodb");

    char str2[] = "cseonefortytwo";
    encryptWord(str2, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str2, "fvhrqhiruwbwzr");

    char str3[] = "the";
    encryptWord(str3, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str3, "wkh");

    char str4[] = "jumps";
    encryptWord(str4, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str4, "mxpsv");

    char str5[] = "a";
    encryptWord(str5, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str5, "d");

    char str6[] = "science";
    encryptWord(str6, "abcdefghijklmnopqrstuvwxyz", 3);
    CuAssertStrEquals(tc, str6, "vflhqfh");

    char str7[] = "cse142";
    encryptWord(str7, "abcdefghijklmnopqrstuvwxyz0123456789", 3);
    CuAssertStrEquals(tc, str7, "fvh475");
    
    char str8[] = "gucci";
    encryptWord(str8, "abcdefghijklmnopqrstuvwxyz0123456789", 3);
    CuAssertStrEquals(tc, str8, "jxffl");
    
    char str9[] = "gang";
    encryptWord(str9, "abcdefghijklmnopqrstuvwxyz0123456789", 3);
    CuAssertStrEquals(tc, str9, "jdqj");
    
    char str10[] = "2";
    encryptWord(str10, "abcdefghijklmnopqrstuvwxyz0123456789", 3);
    CuAssertStrEquals(tc, str10, "5");
    
    char str11[] = "QUICK";
    encryptWord(str11, "ABCDEFGHIJKLMNOPQRSTUVWXYZ", 3);
    CuAssertStrEquals(tc, str11, "TXLFN");
    
    char str12[] = "JUMPS";
    encryptWord(str12, "ABCDEFGHIJKLMNOPQRSTUVWXYZ", 3);
    CuAssertStrEquals(tc, str12, "MXPSV");
    
    char str13[] = "OVER";
    encryptWord(str13, "ABCDEFGHIJKLMNOPQRSTUVWXYZ", 3);
    CuAssertStrEquals(tc, str13, "RYHU");
}

CuSuite* EncryptionMachineGetSuite() {
    CuSuite* suite = CuSuiteNew();
    SUITE_ADD_TEST(suite, testIndexOf);
    SUITE_ADD_TEST(suite, testEncryptLetter);
    SUITE_ADD_TEST(suite, testEncryptWord);
    return suite;
}

void RunAllTests(void) {
    CuString* output = CuStringNew();
    CuSuite* suite = CuSuiteNew();

    CuSuiteAddSuite(suite, EncryptionMachineGetSuite());

    CuSuiteRun(suite);
    CuSuiteSummary(suite, output);
    CuSuiteDetails(suite, output);
    printf("%s\n", output->buffer);
}

int main() {
    RunAllTests();
    return 0;
}


tests/test_suite.c
#include <stdio.h>
#include "CuTest.h"
#include "../EncryptionMachine.h"


void testIndexOf(CuTest* tc) {
    CuAssertIntEquals(tc, indexOf('a'), 0);
    CuAssertIntEquals(tc, indexOf('b'), 1);
    CuAssertIntEquals(tc, indexOf('c'), 2);
    CuAssertIntEquals(tc, indexOf('s'), 18);
    CuAssertIntEquals(tc, indexOf('x'), 23);
    CuAssertIntEquals(tc, indexOf('y'), 24);
    CuAssertIntEquals(tc, indexOf('z'), 25);
}

// encryptLetter and encryptWord were tested with a SHIFT of 3 and an alphabet composed of the 26
// lowercase English letters.
void testEncryptLetter(CuTest* tc) {
    CuAssertIntEquals(tc, encryptLetter('a'), 'd');
    CuAssertIntEquals(tc, encryptLetter('d'), 'g');
    CuAssertIntEquals(tc, encryptLetter('e'), 'h');
    CuAssertIntEquals(tc, encryptLetter('f'), 'i');
    CuAssertIntEquals(tc, encryptLetter('n'), 'q');
    CuAssertIntEquals(tc, encryptLetter('y'), 'b');
    CuAssertIntEquals(tc, encryptLetter('z'), 'c');

}

void testEncryptWord(CuTest* tc) {
    char str1[] = "play";
    encryptWord(str1);
    CuAssertStrEquals(tc, str1, "sodb");

    char str2[] = "cseonefortytwo";
    encryptWord(str2);
    CuAssertStrEquals(tc, str2, "fvhrqhiruwbwzr");

    char str3[] = "the";
    encryptWord(str3);
    CuAssertStrEquals(tc, str3, "wkh");

    char str4[] = "jumps";
    encryptWord(str4);
    CuAssertStrEquals(tc, str4, "mxpsv");

    char str5[] = "a";
    encryptWord(str5);
    CuAssertStrEquals(tc, str5, "d");

    char str6[] = "science";
    encryptWord(str6);
    CuAssertStrEquals(tc, str6, "vflhqfh");
}

CuSuite* EncryptionMachineGetSuite() {
    CuSuite* suite = CuSuiteNew();
    SUITE_ADD_TEST(suite, testIndexOf);
    SUITE_ADD_TEST(suite, testEncryptLetter);
    SUITE_ADD_TEST(suite, testEncryptWord);
    return suite;
}

void RunAllTests(void) {
    CuString* output = CuStringNew();
    CuSuite* suite = CuSuiteNew();

    CuSuiteAddSuite(suite, EncryptionMachineGetSuite());

    CuSuiteRun(suite);
    CuSuiteSummary(suite, output);
    CuSuiteDetails(suite, output);
    printf("%s\n", output->buffer);
}

int main() {
    RunAllTests();
    return 0;
}


src/LWEncryptionMachine.c
#include <stdio.h>
#include <string.h>
#include <inttypes.h>

#include "LWEncryptionMachine.h"


void printIntro(char *ALPHABET, char *SHIFT)
{
    // The longest line is 43 characters long.
    char intro[43] = "Welcome to LWEncryptionMachine";
    puts(intro);
    sprintf(intro, "The program lets you encrypt a message");
    puts(intro);
    sprintf(intro, "with a key for your recipient to decrypt!\n");
    puts(intro);
    
    printf("Alphabet entered: %s\n", ALPHABET);
    printf("Shift value entered: %s\n\n", SHIFT);
}

/// Encrypts a user-entered cryptographic key.
void encryptKey(char *ALPHABET, int SHIFT)
{
    char key[128];
    char encryptedKey[128];

    printf("%s", "Encrypted messages use a shared keyword to decrypt.\n");
    printf("%s", "  Enter key: ");
    fgets(key, 128, stdin);
    
    key[strlen(key) - 1] = '\0';  // Remove the newline character captured by fgets
    strcpy(encryptedKey, key);
    encryptWord(encryptedKey, ALPHABET, SHIFT);
    printf("    \"%s\" has been encrypted to: %s\n\n", key, encryptedKey);
}

/// Encrypts a certain message by using a cipher. The user enters the number of words to be encrypted,
/// and the program encrypts the following user-entered words.
void encryptMessage(char *ALPHABET, int SHIFT)
{
    uint64_t numberOfWords;
    printf("How many words is your message? ");
    scanf("%ld", &numberOfWords);

    // There is a remaining newline character left from stdin that isn't read by stdin. Read that,
    // but don't do anything with it.
    getchar();
    for (int i = numberOfWords; i > 0; --i)
    {
        char word[128];
        char encryptedWord[128];
        printf("  Next word: ");
        fgets(word, 128, stdin);

        word[strlen(word) - 1] = '\0';  // Remove the newline character captured by fgets
        strcpy(encryptedWord, word);
        encryptWord(encryptedWord, ALPHABET, SHIFT);
        printf("    \"%s\" has been encrypted to: %s\n", word, encryptedWord);
    }

    printf("\nMessage fully encrypted. Happy secret messaging!\n");
}

/// Encrypts 1 letter using a Caesar cipher. The letter is encrypted by shifting the letter's position
/// in ALPHABET SHIFT spaces to the right. It basically replaces the letter with another letter further
/// down in the alphabet.
/// @param letter: The char to encrypt
/// @return The encrypted char
char encryptLetter(char letter, char *ALPHABET, int SHIFT)
{
    const int alphabetLength = strlen(ALPHABET);
    const int oldLetterIndex = indexOf(letter, ALPHABET);
    const int newLetterIndex = (oldLetterIndex + SHIFT) % alphabetLength;

    char encryptedLetter = ALPHABET[newLetterIndex];
    return encryptedLetter;
}

/// Encrypt a word using a Caesar cipher.
/// @param word is the string to be encrypted using the cipher. word is modified in-place.
void encryptWord(char word[], char *ALPHABET, int SHIFT)
{
    for (int i = 0; i < strlen(word); ++i)
    {
        word[i] = encryptLetter(word[i], ALPHABET, SHIFT);
    }
}

/// Finds the index of a letter in ALPHABET.
/// @param letter is the char whose index is to be found in ALPHABET.
/// @return The index of letter in ALPHABET, or -1 if the index is not found.
int indexOf(char letter, const char *ALPHABET)
{
    for (int i = 0; i < strlen(ALPHABET); ++i)
    {
        if (letter == ALPHABET[i])
        {
            return i;
        }
    }
    return -1;
}


src/EncryptionMachine.c
#include <stdio.h>
#include <string.h>
#include <inttypes.h>

#include "EncryptionMachine.h"

const char* ALPHABET = "abcdefghijklmnopqrstuvwxyz";
//const char* ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
//const char* ALPHABET = "0123456789";
//const char* ALPHABET = "abcdefghijklmnopqrstuvwxyz0123456789";

#define SHIFT 3


void printIntro() {
    // The longest line is 43 characters long.
    char intro[43] = "Welcome to the CSE142 Encryption Machine";
    puts(intro);
    sprintf(intro, "The program lets you encrypt a message");
    puts(intro);
    sprintf(intro, "with a key for your recipient to decrypt!\n");
    puts(intro);
}

/// Encrypts a user-entered cryptographic key.
void encryptKey() {
	char key[128];
	char encryptedKey[128];

	printf("%s", "Encrypted messages use a shared keyword to decrypt.\n");
	printf("%s", "  Enter key: ");
	fgets(key, 128, stdin);
	
	key[strlen(key) - 1] = '\0';  // Remove the newline character captured by fgets
	strcpy(encryptedKey, key);
	encryptWord(encryptedKey);
	printf("    \"%s\" has been encrypted to: %s\n\n", key, encryptedKey);
}

/// Encrypts a certain message by using a cipher. The user enters the number of words to be encrypted,
/// and the program encrypts the following user-entered words.
void encryptMessage() {
	uint64_t numberOfWords;
	printf("How many words is your message? ");
	scanf("%d", &numberOfWords);

	// There is a remaining newline character left from stdin that isn't read by stdin. Read that,
	// but don't do anything with it.
	getchar();
	for (int i = numberOfWords; i > 0; --i) {
		char word[128];
		char encryptedWord[128];
		printf("  Next word: ");
		fgets(word, 128, stdin);

		word[strlen(word) - 1] = '\0';  // Remove the newline character captured by fgets
		strcpy(encryptedWord, word);
		encryptWord(encryptedWord);
		printf("    \"%s\" has been encrypted to: %s\n", word, encryptedWord);
	}

	printf("\nMessage fully encrypted. Happy secret messaging!\n");
}

/// Encrypts 1 letter using a Caesar cipher. The letter is encrypted by shifting the letter's position
/// in ALPHABET SHIFT spaces to the right. It basically replaces the letter with another letter further
/// down in the alphabet.
/// @param letter: The char to encrypt
/// @return The encrypted char
char encryptLetter(char letter) {
	const int alphabetLength = strlen(ALPHABET);
	const int oldLetterIndex = indexOf(letter);
	const int newLetterIndex = (oldLetterIndex + SHIFT) % alphabetLength;

	char encryptedLetter = ALPHABET[newLetterIndex];
	return encryptedLetter;
}

/// Encrypt a word using a Ceasar cipher.
/// @param word is the string to be encrypted using the cipher. word is modified in-place.
void encryptWord(char word[]) {
	for (int i = 0; i < strlen(word); ++i)
	{
		word[i] = encryptLetter(word[i]);
	}
}

/// Finds the index of a letter in ALPHABET.
/// @param letter is the char whose index is to be found in ALPHABET.
/// @return The index of letter in ALPHABET, or -1 if the index is not found.
int indexOf(char letter) {
	for (int i = 0; i < strlen(ALPHABET); ++i)
	{
		if (letter == ALPHABET[i])
		{
			return i;
		}
	}
	return -1;
}
